realm/realm-library/src/main/cpp/collection_changeset_wrapper.hpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef REALM_JNI_IMPL_COLLECTION_CHANGESET_WRAPPER_HPP
+#define REALM_JNI_IMPL_COLLECTION_CHANGESET_WRAPPER_HPP
+
+#include "collection_notifications.hpp"
+#include "util.hpp"
+#include "jni_util/java_class.hpp"
+#include "jni_util/java_global_weak_ref.hpp"
+#include "jni_util/java_method.hpp"
+#include "jni_util/log.hpp"
+#include "jni_util/jni_utils.hpp"
+#include "jni_util/java_class.hpp"
+#include "jni_util/java_method.hpp"
+#include "sync/partial_sync.hpp"
+
+using namespace realm::jni_util;
+
+namespace realm {
+namespace _impl {
+
+// Wrapper of Object Store CollectionChangeSet
+// It is used to better control the mapping between Object Store concepts and Java API's, especially
+// when it comes to states and defining errors.
+class CollectionChangeSetWrapper {
+public:
+    CollectionChangeSetWrapper(CollectionChangeSet const& changeset, std::string error_message)
+        : m_changeset(changeset)
+        , m_error_message(error_message)
+    {
+    }
+
+    ~CollectionChangeSetWrapper() = default;
+
+    CollectionChangeSetWrapper(CollectionChangeSetWrapper&&) = delete;
+    CollectionChangeSetWrapper& operator=(CollectionChangeSetWrapper&&) = delete;
+    CollectionChangeSetWrapper(CollectionChangeSetWrapper const&) = delete;
+    CollectionChangeSetWrapper& operator=(CollectionChangeSetWrapper const&) = delete;
+
+    CollectionChangeSet& get()
+    {
+        return m_changeset;
+    };
+
+    jthrowable get_error() {
```

Comment:
Return a `JavaGlobalRef` instead to make the life cycle clearer

Reply:
Trying to do this resulted in `JNI DETECTED ERROR IN APPLICATION: use of deleted global`. After spending a few hours on it, I reverted it back to this as it didn't seem important enough to spend more time on.